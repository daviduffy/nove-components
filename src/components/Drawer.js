// External Dependencies
import React from 'react';
import PropTypes from 'prop-types';

const Drawer = ({
  isOpen,
  items,
  label,
  onClick,
  originIndex,
  getInputElement,
  placeholder
}) => (
  <div key="drawer" className="F__g F__g--more">
    {
      label &&
      <span className="F__l"><span>{label}</span></span>
    }
    <div className={`Acrd${isOpen ? ' Acrd--o' : ''}`}>
      <button onClick={onClick} className="Acrd__trg">
        <span className="F__l">
          <span id="drawer_placeholder">{placeholder}</span>
        </span>
      </button>
      <div
        className="Acrd__cont-c"
        style={
          {
            maxHeight: (isOpen && items.length > 0) ? `${(items.length * 110) + 24}px` : 0
          }
        }
      >
        <div className="Acrd__cont">
          {
            items.map((innerInput, i) => getInputElement(innerInput, i, `${originIndex}/items/`))
          }
        </div>
      </div>
    </div>
  </div>
);
Drawer.propTypes = {
  getInputElement: PropTypes.func.isRequired,
  isOpen: PropTypes.bool.isRequired,
  items: PropTypes.array.isRequired,
  label: PropTypes.string,
  onClick: PropTypes.func.isRequired,
  originIndex: PropTypes.string,
  placeholder: PropTypes.string
};
Drawer.defaultProps = {
  label: '',
  originIndex: '',
  placeholder: ''
};
export default Drawer;
