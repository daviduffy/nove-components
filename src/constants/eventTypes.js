export const EVENT_TYPES_COMMON = [
  'ENGAGEMENT',
  'WEDDING',
  'FAMILY',
  'SENIOR'
];

export const EVENT_TYPES_ALL = [
  ...EVENT_TYPES_COMMON,
  'ELOPEMENT',
  'MATERNITY',
  'NEWBORN',
  'COMMERCIAL'
];

export const eventTypes = EVENT_TYPES_COMMON;
