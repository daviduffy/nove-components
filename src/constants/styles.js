export const FORM_STYLES_DEFAULT = {
  borderStyle: 'full',
  borderColor: '#b3b3b3',
  buttonTextColor: '#fefefe',
  custom: '',
  drawerBackgroundColor: '#eeeeee',
  drawerTextColor: '#707070',
  inputBackgroundColor: '#fefefe',
  inputTextColor: '#444444',
  labelTextColor: '#444444',
  negativeColor: '#e74c3c',
  negativeColorBg: '#fadbd8',
  positiveColor: '#2ecc71',
  positiveColorHover: '#2aba67',
  placeholderColor: '#b3b3b3',
  width: '560px'
};

export const FORM_BORDER_PRESETS = [
  // these are all lowercased in the widget so these need to be doubled to prevent UPPERCASING in InputSelect
  ['full', 'full'],
  ['underline', 'underline'],
  ['none', 'none']
];
