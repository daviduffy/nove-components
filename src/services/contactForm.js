import React from 'react';
import PropTypes from 'prop-types';

import Drawer from '../components/Drawer';
import InputDate from '../components/InputDate';
import InputText from '../components/InputText';
import InputSelect from '../components/InputSelect';
import InputTextarea from '../components/InputTextarea';
import Row from '../components/Row';
import Submit from '../components/Submit';

export const getRenderedInputs = ({
  accordionOpen,
  onAccordionClick,
  inputComponents,
  order,
  style
}) => {
  const renderInput = (({ id, items, ...rest }, i) => {
    if (items) {
      let config;
      let Component;
      switch (id) {
        case 'drawer':
          Component = Drawer;
          config = ({
            ...rest,
            key: `${id}-${i}`,
            isOpen: accordionOpen,
            items: items.map(renderInput),
            onClick: onAccordionClick
          });
          break;
        case 'row':
          Component = Row;
          config = ({
            ...rest,
            key: `${id}-${i}`,
            items: items.map(renderInput)
          });
          break;
        default:
          break;
      }
      return (<Component {...config} />);
    }
    if (id === 'submit') return (<Submit key={i} style={style.submit} {...rest} />);
    const { Component, ...REST } = inputComponents.find(({ id: arrayID }) => arrayID === id);
    return (
      <Component key={i} {...REST} />
    );
  });
  renderInput.propTypes = {
    id: PropTypes.string.isRequired,
    items: PropTypes.oneOfType([
      PropTypes.bool,
      PropTypes.array
    ])
  };
  renderInput.defaultProps = { items: false };
  return order.map(renderInput);
};

export const getInputComponents = (props) => {
  const getInputElement = ({
    hidden = false,
    id,
    label,
    options,
    placeholder,
    required,
    type,
    value
  }) => {
    let dynamicOnChange = props.onInputChange;
    let dynamicOnBlur = () => {};
    const config = {
      Component: InputText,
      label,
      id,
      placeholder,
      value,
      required
    };

    switch (type) {
      case 'date':
        return ({
          ...config,
          Component: InputDate,
          error: (required && props.formError && props.eventDate === undefined),
          onDateChange: props.onDateChange,
          value: props.eventDate,
          windowWidth: props.windowWidth
        });
      case 'select':
        config.Component = InputSelect;
        config.options = options ? [['', placeholder], ...options] : false;
        break;
      case 'textarea':
        config.Component = InputTextarea;
        break;
      case 'phone':
        dynamicOnBlur = props.onPhoneBlur;
        break;
      case 'email':
        dynamicOnBlur = props.onEmailBlur;
        break;
      case 'currency':
        dynamicOnChange = props.onAmountChange;
        break;
      default:
        break;
    }

    return ({
      ...config,
      classes: `F__g--${id}`,
      error: required && props.formError !== '' && value === '',
      onChange: dynamicOnChange,
      onBlur: dynamicOnBlur,
      style: ({ ...(hidden ? { display: 'none' } : {}) }),
      required
    });
  };
  return props.inputs.map(getInputElement);
};
