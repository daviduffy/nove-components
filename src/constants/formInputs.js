import { STATES } from './states';
import { EVENT_TYPES_COMMON } from './eventTypes';

export const budget = {
  label: 'Estimated budget',
  placeholder: '$XXXX',
  type: 'currency'
};
export const drawer = {
  label: 'More Details',
  placeholder: 'Share more details about your event here'
};
export const email = {
  label: 'Email',
  placeholder: 'your@email.com',
  type: 'email'
};
export const eventDate = {
  label: 'Event Date',
  placeholder: 'On or around',
  type: 'date'
};
export const eventLocale = {
  label: 'Event Location',
  placeholder: 'Choose a state',
  type: 'select',
  options: STATES
};
export const eventVenue = {
  label: 'Event Venue',
  placeholder: 'Name of the venue'
};
export const name = {
  label: 'Name',
  placeholder: 'Your name'
};
export const message = {
  label: 'Message',
  placeholder: 'What are the details of your event?',
  type: 'textarea'
};
export const phone = {
  label: 'Phone',
  placeholder: '(###) ###-####',
  type: 'phone'
};
export const referralSource = {
  label: 'Referral Source',
  placeholder: 'Where did you hear about me?'
};
export const row = {
};
export const type = {
  label: 'Event Type',
  placeholder: 'Choose an event type',
  type: 'select',
  options: EVENT_TYPES_COMMON
};
export const submit = {
  label: 'Submit'
};

export const FORM_INPUTS_DEFAULT = {
  name,
  email,
  type,
  eventDate,
  referralSource,
  eventVenue,
  eventLocale,
  phone,
  budget,
  drawer,
  row,
  message,
  submit
};

const PRIMITIVE = [
  { id: 'name', path: '0' },
  { id: 'email', path: '1' }
];

const BASE = [
  ...PRIMITIVE,
  { id: 'submit', path: '2' }
];
const MINI = [
  ...PRIMITIVE,
  { id: 'message', path: '2' },
  { id: 'submit', path: '3' }
];
const FULL = [
  ...PRIMITIVE,
  { id: 'type', path: '2' },
  { id: 'eventDate', path: '3' },
  { id: 'referralSource', path: '4' },
  { id: 'drawer', path: '5' },
  { id: 'eventVenue', path: '5/0' },
  { id: 'eventLocale', path: '5/1' },
  { id: 'message', path: '6' },
  { id: 'submit', path: '7' }
];

export const FORM_ORDER = {
  BASE,
  MINI,
  FULL
};
