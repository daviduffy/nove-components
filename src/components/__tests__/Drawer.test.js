// External Dependencies
import React from 'react';
import { shallow } from 'enzyme';

// Internal Dependencies
import Drawer from '../Drawer';
import { FORM_INPUTS_DEFAULT, FORM_ORDER } from '../../constants/formInputs';
import { customDrawerText, stuffedDrawer } from '../../fixtures/forms';
import * as forms from '../../services/contactForm';

// not testing 'remote config' because fetch is performed by SignupRetriever
let wrapper;
const props = {
  isOpen: false,
  ...FORM_INPUTS_DEFAULT.drawer,
  ...FORM_ORDER.FULL[5],
  onClick: () => {}
};

const hydratedInputs = forms.getHydratedInputs({ inputs: FORM_ORDER.FULL });
const renormalizedInputs = forms.renormalizeInputs(hydratedInputs);
const drawer = renormalizedInputs.find(({ id }) => id === 'drawer');

const inputComponents = forms.getInputConfig({
  // this is the flattened inputs with values in SignupContainer
  formError: false,
  onInputChange: () => {},
  inputs: hydratedInputs,
  windowWidth: 1000
});
// console.log(drawer);
const renderedInputs = forms.getRenderedComponents({
  accordionOpen: false,
  inputComponents,
  onAccordionClick: () => {},
  renormalizedInputs: drawer.items, // this is awkward
  style: {}
});

props.items = renderedInputs;

beforeEach(() => {
  wrapper = shallow(<Drawer {...props} />);
});

// drawer
// =================================================================================================
test('should render drawer normally', () => {
  expect(wrapper).toMatchSnapshot();
});

test('should display normal drawer label and placeholder', () => {
  expect(wrapper.find('#drawer_placeholder').text()).toBe(FORM_INPUTS_DEFAULT.drawer.placeholder);
  expect(wrapper.find('.F__g--more > .F__l').text()).toBe(FORM_INPUTS_DEFAULT.drawer.label);
});

test('should display custom drawer placeholder text and no label', () => {
  const renormalizedInputs = forms.renormalizeInputs(customDrawerText.order);
  const drawer = customDrawerText.order.find(({ id }) => id === 'drawer');
  const customProps = { ...props, placeholder: drawer.placeholder, label: '' };
  wrapper = shallow(<Drawer {...customProps} />);
  // should not output label if label set to empty string
  expect(wrapper.find('.F__g--more > .F__l')).toHaveLength(0);
  expect(wrapper.find('#drawer_placeholder').text()).toBe(drawer.placeholder);
});

test('should show items within drawer', () => {
  expect(wrapper.find('.Acrd__cont').children()).toHaveLength(2);
});

test('should set drawer height correctly when multiple items are in it', () => {
  const renormalizedInputs = forms.renormalizeInputs(stuffedDrawer.order);
  const drawer = stuffedDrawer.order.find(({ id }) => id === 'drawer');
  const customProps = { ...props, ...drawer };
  const maxHeightWhenOpen = (drawer.items.length * 110) + 24;
  wrapper = shallow(<Drawer {...customProps} />);
  expect(wrapper.find('.Acrd__cont-c').prop('style')).toEqual({ maxHeight: 0 });
  const customPropsOpen = { ...customProps, isOpen: true };
  wrapper = shallow(<Drawer {...customPropsOpen} />);
  expect(wrapper.find('.Acrd__cont-c').prop('style')).toEqual({ maxHeight: `${maxHeightWhenOpen}px` });
});
