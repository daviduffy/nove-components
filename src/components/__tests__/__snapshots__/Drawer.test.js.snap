// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render drawer normally 1`] = `
<div
  className="F__g F__g--more"
  key="drawer"
>
  <span
    className="F__l"
  >
    <span>
      More Details
    </span>
  </span>
  <div
    className="Acrd"
  >
    <button
      className="Acrd__trg"
      onClick={[Function]}
    >
      <span
        className="F__l"
      >
        <span
          id="drawer_placeholder"
        >
          Share more details about your event here
        </span>
      </span>
    </button>
    <div
      className="Acrd__cont-c"
      style={
        Object {
          "maxHeight": 0,
        }
      }
    >
      <div
        className="Acrd__cont"
      >
        <InputText
          classes="F__g--eventVenue"
          error={false}
          id="eventVenue"
          key="0"
          label="Event Venue"
          onBlur={[Function]}
          onChange={[Function]}
          options={false}
          path="5/0"
          placeholder="Name of the venue"
          required={false}
          style={Object {}}
          value=""
        />
        <InputSelect
          classes="F__g--eventLocale"
          error={false}
          id="eventLocale"
          key="1"
          label="Event Location"
          onBlur={[Function]}
          onChange={[Function]}
          options={
            Array [
              Array [
                "",
                "Choose a state",
              ],
              Array [
                "AL",
                "Alabama",
              ],
              Array [
                "AK",
                "Alaska",
              ],
              Array [
                "AZ",
                "Arizona",
              ],
              Array [
                "AR",
                "Arkansas",
              ],
              Array [
                "CA",
                "California",
              ],
              Array [
                "CO",
                "Colorado",
              ],
              Array [
                "CT",
                "Connecticut",
              ],
              Array [
                "DE",
                "Delaware",
              ],
              Array [
                "FL",
                "Florida",
              ],
              Array [
                "GA",
                "Georgia",
              ],
              Array [
                "HI",
                "Hawaii",
              ],
              Array [
                "ID",
                "Idaho",
              ],
              Array [
                "IL",
                "Illinois",
              ],
              Array [
                "IN",
                "Indiana",
              ],
              Array [
                "IA",
                "Iowa",
              ],
              Array [
                "KS",
                "Kansas",
              ],
              Array [
                "KY",
                "Kentucky",
              ],
              Array [
                "LA",
                "Louisiana",
              ],
              Array [
                "ME",
                "Maine",
              ],
              Array [
                "MD",
                "Maryland",
              ],
              Array [
                "MA",
                "Massachusetts",
              ],
              Array [
                "MI",
                "Michigan",
              ],
              Array [
                "MN",
                "Minnesota",
              ],
              Array [
                "MS",
                "Mississippi",
              ],
              Array [
                "MO",
                "Missouri",
              ],
              Array [
                "MT",
                "Montana",
              ],
              Array [
                "NE",
                "Nebraska",
              ],
              Array [
                "NV",
                "Nevada",
              ],
              Array [
                "NH",
                "New Hampshire",
              ],
              Array [
                "NJ",
                "New Jersey",
              ],
              Array [
                "NM",
                "New Mexico",
              ],
              Array [
                "NY",
                "New York",
              ],
              Array [
                "NC",
                "North Carolina",
              ],
              Array [
                "ND",
                "North Dakota",
              ],
              Array [
                "OH",
                "Ohio",
              ],
              Array [
                "OK",
                "Oklahoma",
              ],
              Array [
                "OR",
                "Oregon",
              ],
              Array [
                "PA",
                "Pennsylvania",
              ],
              Array [
                "RI",
                "Rhode Island",
              ],
              Array [
                "SC",
                "South Carolina",
              ],
              Array [
                "SD",
                "South Dakota",
              ],
              Array [
                "TN",
                "Tennessee",
              ],
              Array [
                "TX",
                "Texas",
              ],
              Array [
                "UT",
                "Utah",
              ],
              Array [
                "VT",
                "Vermont",
              ],
              Array [
                "VA",
                "Virginia",
              ],
              Array [
                "WA",
                "Washington",
              ],
              Array [
                "WV",
                "West Virginia",
              ],
              Array [
                "WI",
                "Wisconsin",
              ],
              Array [
                "WY",
                "Wyoming",
              ],
            ]
          }
          path="5/1"
          placeholder="Choose a state"
          required={false}
          style={Object {}}
          value=""
        />
      </div>
    </div>
  </div>
</div>
`;
